[RFC] Fix clang-scan issues

sander2 opened the pull request:
Maybe we should change branch names, since we do not fix dead stores anymore.

As a side note, there are about 300 other occurrences of if statements that use `has_mbyte` that we could remove..

It would be nice to fix all issues in https://neovim.io/doc/reports/clang/ :
- [ ] Dead assignment	in eval/typval_encode.c.h, function encode_vim_to_object on line 716
- [ ] Dead assignment	in eval/typval_encode.c.h, function encode_vim_to_string on line 716
- [ ] Dead assignment	in eval/typval_encode.c.h, function encode_vim_to_echo on line 716
- [ ] Dead assignment	in eval/typval_encode.c.h, function encode_vim_to_json on line 716
- [ ] Dead assignment	in eval/typval_encode.c.h, function encode_vim_to_msgpack on line 716
- [x] Dead assignment	in hardcopy.c, function hardcopy_line on line 848
  - [x] discussed
- [x] Dead assignment	in message.c, function trunc_string on line 273
  - [x] discussed
- [x] Dead initialization	in misc1.c, function ins_char_bytes on line 1408
  - [x] discussed
- [ ] Dead assignment	in screen.c, function win_line on line 3077
  * Probably shouldn't touch this one, as it is inside a 2000 line function filled with inconsistencies. It will probably not be appreciated if we fix a small part of it and leave the rest alone.
- [x] Dead increment	in screen.c, function win_redr_status on line 4879
  - [x] discussed
- [x] Dead assignment	in strings.c, function vim_vsnprintf on line 982
  - [x] discussed

Hjdskes committed:
screen.c: account for translated string length

`[RO]` is appended to the status line and `len` is increased with the length of
this string (4). However, the string is marked for translation and may
thus well be larger (or smaller) than 4. Therefore, we check the length at
runtime. The resulting len is never actually used, and thus could be removed.
However, by keeping this line, the body of this if-statement is kept consistent
with surrounding code, and future changes can not forget to add this line when
additional strings are added to p.

sander2 committed:
screen.c: remove dead code

As stated in globals.h, mbyte flags are deprecated, and code using it can be
refractored to remove dead code. Since has_mbyte is defined to true, this
refractoring correct.

Hjdskes commented:
Nice, finally a PR to really discuss our commits.

Some comments:

1. The name of this PR should be related to the changes made within it. I propose changing it to something along the lines of *Fix clang-scan issues*.
2. We should also add a list of remaining items in that list to the first post in this PR, so we can simply tick those off once the related commit has been pushed/discussed and keep progress, e.g.

- [ ] Dead assignment in `eval/typval_encode.c.h`, function `encode_vim_to_object` on line 716, [report](https://neovim.io/doc/reports/clang/report-3d7d52.html#EndPath).
3. We should add a [WIP] tag to the title, e.g. `[WIP] <title>`.

> Maybe we should change branch names, since we do not fix dead stores anymore.

Can we do this while leaving this PR intact? If so, I'm all for it.

Hjdskes commented:
Regarding the commit messages, Neovim guidelines say they want to have the imperative voice: https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#commit-messages

@sander2 Can you change your commit to reflect this?

sander2 commented:
I made the changes you suggested. Also, I was under the impression that we would only fix one issue. Now, after discussing this with you, I agree that we can try to fix them all, if this is doable without investing too much time. In that case, the branch name is correct and we don't need to change it.

sander2 committed:
misc1.c: remove dead initialization

sander2 committed:
message.c: fix dead assignment by removing dead code

`enc_dbcs` and `enc_utf8` are deprecated (globals.h), so the second branch is
always taken.

sander2 committed:
hardcopy.c: fix dead assignment

`has_mbyte` is deprecated (globals.h), so `outputlen` is always assigned
within the if statement. Therefore, the previous initialization is
unnecessary.

Hjdskes commented on src/nvim/misc1.c:
```diff
@@ -1498,10 +1497,7 @@ void ins_char_bytes(char_u *buf, size_t charlen)
       && !ins_compl_active()
       ) {
-    if (has_mbyte)
       showmatch(mb_ptr2char(buf));
```
The indentation needs to be two spaces less here now.

Hjdskes commented on src/nvim/message.c:
```diff
@@ -270,32 +270,16 @@ void trunc_string(char_u *s, char_u *buf, int room, int buflen)
-      i += (*mb_ptr2len)(s + i);
-    }
-  } else if (enc_utf8) {
-    /* For UTF-8 we can go backwards easily. */
```
Maybe this comment should be kept, to indicate what is going on (i.e., going backwards)

sander2 replied:
I considered it, but decided not to, as from the other comment (` For DBCS going backwards in a string is slow[...]`), I understood that backwards is sort of the default direction. The comment you want to keep is more of a contrast to the previous comment; it is not really an independent comment imo. If we want to keep it, I think we should change it to something like 'iterate backwards over the strings; this is reasonable fast for UTF-8'.

Hjdskes replied:
> If we want to keep it, I think we should change it to something like 'iterate backwards over the strings; this is reasonable fast for UTF-8'.

That was more along the lines of what I meant (possible typo: reasonably)

Hjdskes commented on src/nvim/message.c:
```diff
@@ -270,32 +270,16 @@ void trunc_string(char_u *s, char_u *buf, int room, int buflen)
-      len += n;
+  half = i = (int)STRLEN(s);
+  for (;; ) {
+    do
```
You should apply brackets here according to the style guide.

Hjdskes commented on src/nvim/message.c:
```diff
@@ -270,32 +270,16 @@ void trunc_string(char_u *s, char_u *buf, int room, int buflen)
-    for (i = (int)STRLEN(s); len + (n = ptr2cells(s + i - 1)) <= room; --i)
-      len += n;
+  half = i = (int)STRLEN(s);
+  for (;; ) {
```
Style guide: remove space (I think)

Hjdskes commented on src/nvim/message.c:
```diff
@@ -270,32 +270,16 @@ void trunc_string(char_u *s, char_u *buf, int room, int buflen)
-  } else {
-    for (i = (int)STRLEN(s); len + (n = ptr2cells(s + i - 1)) <= room; --i)
-      len += n;
+  half = i = (int)STRLEN(s);
```
Style guide: is an assignment like this allowed?

Hjdskes commented on src/nvim/message.c:
```diff
@@ -270,32 +270,16 @@ void trunc_string(char_u *s, char_u *buf, int room, int buflen)
+      half = half - (*mb_head_off)(s, s + half - 1) - 1;
+    while (utf_iscomposing(utf_ptr2char(s + half)) && half > 0);
+    n = ptr2cells(s + half);
+    if (len + n > room)
```
Style guide: brackets

sander2 commented:
You mention a couple of style changes. Are you sure we should apply stylistic changes, even though this isn't the focus of the commits?

Hjdskes commented:
> You mention a couple of style changes. Are you sure we should apply stylistic changes,  even though this isn't the focus of the commits?

Yes, this is mandatory. Changes that do not conform to the style guide are not accepted.

sander2 commented:
That makes sense, I suppose. Should we put all stylistic changes into a single separate commit, or can we just add them to the commits we already made? 

By the way (mostly note to self), if we are to make stylistic changes, then we need to change all comments to `//` style as well.

Hjdskes commented:
> That makes sense, I suppose. Should we put all stylistic changes into a single separate commit, or can we just add them to the commits we already made?

They should be in the same commit as the actual change you're making.

> By the way (mostly note to self), if we are to make stylistic changes, then we need to change all comments to // style as well.

Yes, we should all read the style guide and/or use `make lint` or `make clint` after we are done with our changes, see https://github.com/neovim/neovim/wiki/Development-tips#code-linting

Hjdskes committed:
screen.c: remove unused assignment

As reported by clang-scan, `length_modifier` is never read in any code path
following this branch. It is safe to remove.

sander2 commented on src/nvim/strings.c:
```diff
@@ -979,7 +979,6 @@ int vim_vsnprintf(char *str, size_t str_m, const char *fmt, va_list ap,
           const void *ptr_arg = NULL;
 
           if (fmt_spec == 'p') {
-            length_modifier = '\0';
```
I agree that it should be safe to remove. This code is within a while-loop, but `length_modifier` is declared and initialized within the loop, so only the code below this statement needs to be checked.

sander2 reviewed (commented):
(testing review feature of github, don't mind this comment for now..)

Hjdskes commented:
@sander2 You should probably fix the style of the whole function you work in, there aren't many lines in there that you haven't touched already.

sander2 committed:
lint: fix lint warnings

sander2 committed:
Merge branch 'clang-scan-fix-dead-stores-dev' of https://github.com/delftswa2017/neovim into clang-scan-fix-dead-stores-dev

Hjdskes commented:
All changes have been reviewed by at least @sander2 and myself. `make lint` does not report any errors in code we have touched, and `make test` fails the same tests that currently fail on master (due to an indeterminism in code we did not touch) so that is probably alright too.

Some questions we have to ask the upstream maintainers:
1. How much of the surrounding code are we supposed to update to the new style?
2. Are there any tests for code we have changed, or do we have to add tests for changes we made? If so, can you point us to where these tests should go?
3. The fix made in the first commit is also found in upstream vim (https://github.com/vim/vim/blob/master/src/screen.c#L6732); we should tell the Neovim maintainers this so they can label our PR as such. They might ask us to make a PR to vim as well.
4. Related to the first commit, we should ask the Neovim developers if we made the right choice by leaving `len` in the last branch.

We are still looking at the issues outstanding in `eval`; it seems these `#define`s are redefined in files that include this file and so these are false positives.

@Saminobi @TMillross It would be great if you could review the changes as well! I will merge this and open a PR to Neovim either during or right after our meeting tonight.
